Complete the doc of the XKB keymap text format, V1

wismill opened the pull request:
XKB text format is really tricky and documentation is quite scarce. There are well known sources (see current doc for references), but there is a need for a complete documentation of the current implementation of this format here.

The only way to know how some features work it to look at the code.

So I have the ambition to complete this documentation with the following approach:
- Prefer direct actionable specification with examples than a formal one.
- Document only features implemented by `xkbcommon`, i.e. “modern XKB”. When possible, document differences with the features supported by X11 and the choices that may look arbitrary at first sight (e.g. keycodes starting at 8).
- Mark known missing documentation explicitly so. It avoids user to miss features or limitations and facilitates the documentation work.

__NOTE:__ unfortunately the Markdown format supported by `doxygen`  is not supported by Github. In order to have proper rendering, please build the documention and use Github online rendering in last as a last resort.

wismill commented:
@whot I would like to know your opinion on how to proceed:

- First do you agree on the approach?
- What do you think of the first changes in the structure?
- Should I use only one commit per section, or a single commit, a PR per section, or something else?

Note: I have no timeline for this. Maybe it would be better to have a merge request per milestone.

bluetech commented on doc/keymap-format-text-v1.md:
```diff
@@ -461,12 +809,398 @@ statement only defines the Group3 of a mapping:
+- `LatchGroup`: TODO
+- `LockGroup`: TODO
+
+### Pointer actions
```
All actions from here on are not supported by xkbcommon, i.e. they are parsed and validated but have no effect.

wismill replied:
Fixed

bluetech reviewed (commented):
Thanks @wismill, this is excellent.

I've read everything; though I didn't triple check every statement, it seems correct to me. Only left some minor comments.

IMO we can merge this already, and I'll be happy to review any further PRs you send. Let me know if you want to do so, or keep working in this PR.

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -3,21 +3,326 @@
+  <dd>
+  A mapping of keycodes to symbols, actions and key types.
+
+  A user that deal with multiple languages may need two or more
```
Done

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -3,21 +3,326 @@
 This is the simplest section type, and is the first one to be
 compiled. The purpose of this is mostly to map between the
-hardware/evdev scancodes and xkb keycodes. Each key is given a name
+hardware/evdev scancodes and XKB keycodes. Each key is given a name
```
Done

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -31,21 +336,46 @@ The above would let 49 and 10 be valid keycodes in the keymap, and
 
+Note that contrary to xkbcommon, the X11 protocol supports keycodes
+only up to `255`. Keymaps and applications which are compatible with
+X11 should _not_ use keycodes beyond `255`.
```
Fixed

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -3,21 +3,326 @@
+
+  See [xkb_keymap] for further details.
+  </dd>
+</dl>
```
Done

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -73,7 +403,7 @@ index. The LED may be referred by this name later in the compat
 
 
-## The `xkb_types` section
+## The `xkb_types` section {#the-xkb_types-section}
```
Yeah, doxygen flaws… It must be plain text. Fixed

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -461,12 +809,398 @@ statement only defines the Group3 of a mapping:
+(see `XkbNumVirtualMods`).
+
+The following table summarizes the modifiers defined
+in <code>[xkeyboard-config]</code>.
```
Fixed

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -461,12 +809,398 @@ statement only defines the Group3 of a mapping:
+    virtual_modifiers LevelThree;
+
+    type "FOUR_LEVEL" {
+        // Key type modifier mask: all the modifiers
```
Fixed

wismill commented on doc/keymap-format-text-v1.md:
```diff
@@ -461,12 +809,398 @@ statement only defines the Group3 of a mapping:
+    // we do not need to add it here.
+    virtual_modifiers LevelThree;
+
+    // Set defaults. They are overriden if the set directly
```
Fixed

wismill commented:
@whot I fixed the issue and made some improvements. 

I added the keysyms header to the doc, is it OK? It enable a nice hyperlink from the format doc page.

> IMO we can merge this already, and I'll be happy to review any further PRs you send. Let me know if you want to do so, or keep working in this PR.

Multiple PR are much better! I am done with this one.

bluetech commented:
Thanks again. Merging now.

> I added the keysyms header to the doc, is it OK? It enable a nice hyperlink from the format doc page.

It does create a 1.2MB html, but that's not so bad and I see the benefit.
